### ISSUE TITLE
Lagoon Implementation Discovery 

### ISSUE THREAD

**Issue by olivereri** [department-of-veterans-affairs/va.gov-cms#13113]
## Description
Building on the work previously done by @ndouglas #6674 we seek define any pitfalls, difficulties or unique requirements to implementing Lagoon based on its requirements. This should as much as possible be agnostic of the proposed architecture. For example, exploring EKS access directly to pods running Lagoon or our application.

## Acceptance Criteria
- [ ] Lagoon installation requirements are defined and compared against available resources
- [ ] Testable_Outcome_Y
- [ ] Testable_Outcome_Z
- [ ] Requires design review

### Team
Please check the team(s) that will do this work.

- [x] `CMS Team`
- [ ] `Public Websites`
- [ ] `Facilities`
- [ ] `User support`

**Comment by ElijahLynn** [comment]
Here is our living Lagoon SuperEpic of sorts > https://docs.google.com/document/d/1GmgzPOqcQz9RUnjUnfuEK_5_NRRAndoVTYPIKXIUHNI/edit#

**Comment by ndouglas** [comment]
Created #13253, #13254, #13255, #13256, #13263, #13264, #13265, #13266, #13267, and #13268 (and commented on #13163) based on review of a video conference between Elijah and Eric and Amazee.io representatives. 

I'll add more issues based on my braindump later:

* Discovery: discovery, analysis, identification, evaluation. Mostly complete but not really documented outside of the RfC and scattered issues/Slack convos/video conferences/markdown-formatted existential crises.
   * Identify feature deficits in Lagoon we would like to address. Open followup issues to prioritize and address the particulars. These are likely going to be theoretical or conjectural in nature.
   * Investigate the existing deployments, e.g. in the Australian government. What issues have they encountered? Have they managed to address these issues? Do these apply to us? If so, open followup issues to prioritize and address. These are probably going to be practical, real-world in nature.
   * Identify, research, and implement prerequisite systems and dependencies, and open up followup issues. For instance, Nexus needs to be configured to support Drupal. Lagoon does not like mouf/node-installer, so we need to move to `n` or `nvm`. What other changes need to be made?
   * Plan and conduct a security evaluation. We performed a similar process for Tugboat; what will we need to do here? What procedure can we follow?
   * Plan and conduct Lagoon demonstrations and training. This should occur for both DevOps engineers as well as limited, small-group demos and training performed by DevOps engineers and targeted at other engineers and team members. This will boost everyone’s knowledge and expose gaps in the DevOps engineers’ understanding of the system.
   * Write a retrospective of the discovery process spanning the past year or whatever. This has been a scattered, stop-and-start process, and we should collect/summarize/document the process in a single place so that we can refer to it as needed.
      * Refer to:
         * original ArgoCD vs. Lagoon docs
         * original Kubernetes learning issues
         * RFC
         * videos (I need to watch and rewatch these)
         * Work done elsewhere in this discovery process
      * Include:
         * Summary of BRD infrastructure state, including tools and processes
         * Pressures to move off BRD
         * Platform Ops option and why we are like “meh”
         * Lagoon option
         * Other considered options (the initially-considered ones)
* Justification: RfC is submitted, but we may need to defend it or appeal or find alternative methods of funding. I don’t have a great understanding of this process.
   * Defend and revise the RfC; we have an open PR, but it will need review and any comments raised will need to be addressed. Once it is merged, we can move forward.
   * Get signed support contract. This should ideally include:
      * Additional development hours from Amazee.io engineers to devote to remedying feature deficits identified during the discovery phase.
      * DDEV support for Lagoon developed by DDEV engineers
   * Identify funding sources and alternative options if necessary. We should consider the possibility that Amazee.io engineering hours and/or DDEV engineering hours will not be paid and we will need to perform this labor ourselves.
   * #13111 is aimed at providing an exhaustive level of transparency and insight into the system as currently planned.
* Documentation: We don’t have sufficient comprehensible diagrams or other documentation covering the new architecture. We should produce these artifacts in preparation for the developer rollout.
   * See #13111. These resources can be refined on an ongoing basis to reflect changes and improvements in our understanding, and share knowledge not just with the customer but with the larger team.
      * Lagoon Core architecture diagrams/documentation
      * Lagoon Remote architecture diagrams/documentation
   * Develop a plan for documenting the system. Scope, objectives, etc.
   * Write specific READMEs for the repo (e.g. replacing brd.md, updating local.md, etc).
   * Plan and deliver a documentation information session. This could be a quick knowledge share. We just want to publicize that we have this documentation available, for developers as well as the customer.
* Planning: actual migration plan - timebook, resources, milestones, etc. How we manage switchover for editors. Rollback plan, because we could conceivably discover a showstopper once the switchover has occurred.
   * Outline the migration process - timelines, resource allocation, roles and responsibilities. This needs to include all relevant personnel, where issues will be communicated if they arise, etc. Also need to have all necessary credentials and tools, e.g. if we’re going to need to have SSH access to troubleshoot something, at what stage is that available? How do we ensure that we do not lose insight or control? Also include risk management plans, identify potential risks and strategies for mitigating them.
   * Define migration milestones - identify key increments in the process, such as completing the dev environment setup, staging migration, production switchover, etc
   * Define the switchover plan for editors - I don’t think we need training here, but we should plan communications for:
      * Announcing some time in advance (“Next week we’re gonna do the thing”)
      * Announcing at the time of (“We’re doing the thing, hold on to yer butts”)
      * Announcing successful completion (“We did the thing, pls let us know if anything breaks”)
      * Announcing failure (“Uh, something unexpected happened and we’re rolling back. Should be okay in x minutes.”)
   * Develop a rollback plan - Write this in tandem with the migration process. Is each step reversible? What’s the point of no return? What’s the estimated time needed to reverse course or undo from each point? Can we provide an ELI5 explanation for each so we can quickly communicate the rollback to stakeholders while reassuring them that normal services will be restored quickly?
   * Plan decommissioning of BRD - What is the threshold at which we move forward on decommissioning BRD? A day? A week? A sprint? When do we assume that the new system is stable and we can remove the old? What does this process involve? What risks does this involve?
   * Schedule training for other team members - Need to review when/where this is applicable, who’s performing the training, etc.
* Dev/Staging Validation: we need to track errors, etc, ensure that our logging is functioning as intended, Sentry is doing things, DataDog is doing things, etc. Then we can do some chaos monkey stuff, mob violence, etc to try and uncover new issues. Delicate period since we can’t assume passing tests mean everything’s great – we need to force errors and challenge our assumptions to ensure reporting is accurate.
   * Set up and configure logging and monitoring integrations - ensure they’re functioning as intended. Spoof messages to verify delivery where possible. If not, demonstrate availability and reliability some other way.
      * Sentry?
      * DataDog
      * In-house things?
      * Where do EKS logs go?
      * Are there other logs, e.g. application logs?
      * Accommodate the known issue of log visibility in Lagoon UI, which TBF I don’t know the full details of.
   * Test error tracking and reporting - intentionally generate errors in dev/staging to validate systems. Create a list of errors to generate, e.g. PHP startup errors vs. PHP runtime errors, PHP runtime notices, etc all have different implications, will be handled differently, and thus need to be verified differently. 
   * We do a little trolling - chaos monkey test the environment. Can this be done with Cypress? What about other automated tools? See what comes up, and document them as they arise.
   * Perform load and stress testing - simulate high load, e.g. with Siege or some other tool, and see how scaling operates in practice and whether it behaves as intended. Document discrepancies or concerns as they arise.
   * Refine and iterate - issues need to be prioritized accordingly and may need to be resolved prior to migration. Others may be treated as “known issues”.
* Switchover: issues actually related to live migration of prod. Schedule a maintenance window for the migration, execute the plan, monitor the process, address issues that arise. This includes ops changes like DNS records, etc.
   * Schedule a maintenance window - Check scheduled PTO, check personnel health, make sure everyone’s ready and alert. Check with stakeholders and other teams, possibly, to confirm that we’re minimizing the potential disruption to users and services.
   * Carry out the runbook as written, monitoring and addressing and announcing and so forth as needed.
   * Verify that everything has been migrated successfully. Testing, etc. If serious issues arise at this point, we may need to rollback. Analyze and verify performance under real-world conditions. Does this match expectations?
* Decommission BRD: Once we’re confident that Lagoon is behaving as expected, we need to decommission BRD gracefully. Possible risks should be identified, a schedule created, announcements made, etc. This, like Switchover, should follow the plan established during the Planning phase.
* Project Retrospective: We should write a retrospective of the entire project and how we planned and executed it. It might be useful to Amazee.io, stakeholders, internal documentation for future initiatives, etc.
   * Collect project data and metrics – how long did this take, beginning to end? How long did specific phases take? What resources did it occupy? What issues were encountered? What decisions were made, and at what time?
   * Identify successes and achievements - what did we do right? What was smooth?
   * Analyze self-identified areas for improvement - What could’ve been done better? Where did communication or processes break down? Where did we have to switch gears or horses? What systemic things contributed to or caused these issues? What would we recommend? We should have a meeting about this where the principal team members discuss this.
   * Gather feedback from outside team members and stakeholders - How did they feel about the process? Was it transparent? Responsible? Accountable? What deficits remain?
   * Create action plan for followups - Ensure that every deficit in the retrospective report has a corresponding issue or AC somewhere to address and resolve it.
   * Complete the retrospective report - Just write all this out in some coherent fashion.
   * Distribute the retrospective report - To Amazee.io, stakeholders, ToT(?), put it in the repo, etc.
